### PR TITLE
Answer

### PR DESCRIPTION
**Pull request by geoffrey112**

### CODE REVIEW

File: Functions/README.md [hunk 1]
```diff
@@ -14,7 +14,7 @@ Methods ! here we gonna study how to build functions with or without parameters.
     - func() == 'MAC OS X'
 
 #### With one parameter
-1. Create a function which take a number as parameters and return and return 'true' if it is a pair number and 'false' when it is odd 
+1. Create a function which take a number as parameters and return 'true' if it is a pair number and 'false' when it is odd
```
Comment (wehrli) [inline Functions/README.md:17]: Thank you 🙏

File: Functions/src/js/func.js [hunk 2]
```diff
@@ -0,0 +1,146 @@
+// Without parameters 
+// 1.
+function myDate(){
+  let date = new Date();
+  return date;
```
Comment (wehrli) [inline Functions/src/js/func.js]: Here you return the date non-formatted.
**you got**: `Tue Nov 17 2020 15:30:00 GMT+0000 (Greenwich Mean Time)`
**expected**: `Tuesday 17 November 2020, 3:30pm`

Click [here](https://www.w3schools.com/js/js_date_formats.asp) 😉

File: Functions/src/js/func.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+
+// 2.
+function myOs(){
+  let OS = "OS Inconnu"; 
+  if (navigator.appVersion.indexOf("Win")!=-1) OS = "Your operating system is: Windows"; 
+  if (navigator.appVersion.indexOf("Mac")!=-1) OS = "Your operating system is: MacOS"; 
+  if (navigator.appVersion.indexOf("X11")!=-1) OS = "Your operating system is: Unix"; 
+  if (navigator.appVersion.indexOf("Linux")!=-1) OS = "Your operating system is: Linux";
+  return OS;
+}
```
Comment (wehrli) [inline Functions/src/js/func.js]: nice use of `navigator.appVersion`, this was the solution ;)

For a better optimisation you could have use an [enum and a switch](https://www.sohamkamani.com/blog/2017/08/21/enums-in-javascript/)

File: Functions/src/js/func.js [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+
+
+// 2.
+function returnPI(nb){
+  let pi = Math.PI;
+  let virgule = pi.toFixed(nb);
+  return virgule;
+
+  // return Math.PI.toFixed(nb); // Other answer
```
Comment (wehrli) [inline Functions/src/js/func.js]: both technique works, but one is more readable than the other. 😉

File: Functions/src/js/func.js [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+document.getElementById("device").innerHTML = myOs();
+
+
+////////////////////////////////////
+
+
+// With one parameter 
+// 1.
+function checkNumber(nb){
```
Comment (wehrli) [inline Functions/src/js/func.js]: 🥇

File: Functions/src/js/func.js [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+  // string = string.split(""); 
+  // string = string.reverse(); // Second answer found
+  // string = string.join("");
+  // return string;
+
+
+  // return string.split("").reverse().join(""); //Other smaller answer seen on internet 
+
+  // But what is the best answer?
```
Comment (wehrli) [inline Functions/src/js/func.js]: which one it is according to you ? 🤔

File: Functions/src/js/func.js [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+  return string.length;
+}
+
+document.getElementById("array").innerHTML = lengthArray(["M4 ", "G18 ", "AWP"]);
+
+
+
+// 6
+function returnArray(string){
+  return string.split(" ");
```
Comment (wehrli) [inline Functions/src/js/func.js]: a bit too static according to me, what if we want to accept multiple separator? 

[regexp](https://developer.mozilla.org/fr/docs/Web/JavaScript/Reference/Objets_globaux/RegExp)

File: Functions/src/js/func.js [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+document.getElementById("returnArray").innerHTML = returnArray("Hello world!");
+
+
+
+// 7.
+function returnObject(string){
```
Comment (wehrli) [inline Functions/src/js/func.js:170]: you fell into my trap 😈 

read [that](https://www.w3schools.com/js/js_json_intro.asp) first
and [that](https://www.w3schools.com/js/js_json_parse.asp), very important as a web developper

File: Functions/src/js/func.js [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+document.getElementById("remove").innerHTML = removeSec("Hello world !", "world");
+
+
+// Test without function and it works
+
+// let txt = "Hello world !";
+// let replace = txt.replace("world ", "");
+
+// console.log(replace);
```
Comment (wehrli) [inline Functions/src/js/func.js]: Look at you code deeply, read it out loud. the answer is under your eyes 🔍

**Review by wehrli — CHANGES_REQUESTED** [review]
It is a very good start, you manage to use the javascript library ;) well use of `includes` & `toFixed`.
You hesitate sometimes about how to code it. Well I guess here it is just a matter of how you like it. There is debate on that anyway. I personally prefer to be easy to read. I can show you at the meeting ;) 
Nice implementation of the result through the html, I just regret that i had no way to put new value from it 😞

Most of your methods are not protected against unwanted value, like if put an `int` as parameters, rather than a `string`. YOU MUST ALWAYS VERIFY IT IS A WANTED VALUE

File: Functions/src/js/func.js [hunk 10]
```diff
@@ -1,8 +1,81 @@
 // Without parameters 
 // 1.
 function myDate(){
-  let date = new Date();
-  return date;
+  let day = new Date().getDay();
```
Comment (wehrli) [inline Functions/src/js/func.js:4]: In here it is a working algorithm nice ! 

now you could have use more the Date library to do it in 2 or three line (aka `toLocaleDateString` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toLocaleDateString)

Also you are creating new Date everytime, you could have done
```
const date = new Date();
const day = date.getDay();
const month = date..getMonth();
....
```

File: Functions/src/js/func.js [hunk 11]
```diff
@@ -27,73 +101,68 @@ document.getElementById("device").innerHTML = myOs();
...
 // With one parameter 
 // 1.
-function checkNumber(nb){
-  return nb % 2 == false;
-}
 
-document.getElementById("checkNb").innerHTML = checkNumber(4);
+let button1 = document.getElementById("checkNumber");
+let input1 = document.getElementById("checkNumberInput");
+let invalidChars = ["e", "E", "+", ",", "."];
```
Comment (wehrli) [inline Functions/src/js/func.js:107]: why the letter `e` is invalid char ?

File: Functions/src/js/func.js [hunk 12]
```diff
@@ -27,73 +101,68 @@ document.getElementById("device").innerHTML = myOs();
...
+
+input1.addEventListener('keydown', function(chars){
+  if (invalidChars.includes(chars.key)) {
+    chars.preventDefault();
+  }
+});
+
+button1.addEventListener('click', function(nb){
+  nb = document.getElementById("checkNumberInput").value;
+  let answer = document.getElementById("checkNb").innerHTML = nb % 2 == false;
```
Comment (wehrli) [inline Functions/src/js/func.js:117]: no need to create the variable answer here if you are not using it ;)

File: Functions/src/js/func.js [hunk 13]
```diff
@@ -27,73 +101,68 @@ document.getElementById("device").innerHTML = myOs();
...
-  // return string.split("").reverse().join(""); //Other smaller answer seen on internet 
-
-  // But what is the best answer?
-}
-
-document.getElementById("reverse").innerHTML = reverse("Hello world!");
+let button3 = document.getElementById("reverseButton");
+button3.addEventListener('click', function(input){
+  input = document.getElementById("inputReverse").value;
+  let answer = document.getElementById("reverse").innerHTML = input.split("").reverse().join("");
```
Comment (wehrli) [inline Functions/src/js/func.js:135]: same here no need of the variable answer

File: Functions/src/js/func.js [hunk 14]
```diff
@@ -27,73 +101,68 @@ document.getElementById("device").innerHTML = myOs();
...
-function lengthArray(string){
-  return string.length;
-}
-
-document.getElementById("array").innerHTML = lengthArray(["M4 ", "G18 ", "AWP"]);
+let button5 = document.getElementById("lengthArray");
+button5.addEventListener('click', function(array){
+  array = document.getElementById("inputArray").value.split(/[,]/g);
+  let answer = document.getElementById("array").innerHTML = array.length;
+  console.log(array);
```
Comment (wehrli) [inline Functions/src/js/func.js:154]: leftover :p

File: Functions/src/js/func.js [hunk 15]
```diff
@@ -27,73 +101,68 @@ document.getElementById("device").innerHTML = myOs();
...
 
 // 5.
-function lengthArray(string){
-  return string.length;
-}
-
-document.getElementById("array").innerHTML = lengthArray(["M4 ", "G18 ", "AWP"]);
+let button5 = document.getElementById("lengthArray");
+button5.addEventListener('click', function(array){
+  array = document.getElementById("inputArray").value.split(/[,]/g);
```
Comment (wehrli) [inline Functions/src/js/func.js:152]: `/[,]/` beggining of using regexp ! nice !

File: Functions/src/js/func.js [hunk 16]
```diff
@@ -27,73 +101,68 @@ document.getElementById("device").innerHTML = myOs();
...
-function returnArray(string){
-  return string.split(" ");
-}
-
-document.getElementById("returnArray").innerHTML = returnArray("Hello world!");
+let button6 = document.getElementById("ArrayString");
+button6.addEventListener('click', function(string){
+  string = document.getElementById("inputString").value;
+  let answer = document.getElementById("arrayString").innerHTML = string.split(" ");
+  console.log(string.split(" "));
```
Comment (wehrli) [inline Functions/src/js/func.js:164]: leftover :p

File: Functions/src/views/index.html [hunk 17]
```diff
@@ -11,43 +11,128 @@
...
+            <p id="reverse"></p>
+          </div>
+        </li>
+
+        <!-- Exercice 4 -->
+        <li>
+          <div class="input-group mt-2">
+            <input type="text" id="inputLength" class="form-control" placeholder="Enter string" aria-label="Length of string" aria-describedby="Length of string">
+            <div class="input-group-prepend mr-2">
+              <button class="btn btn-outline-primary btn-sm" type="button" id="checkLength">Lenght</button>
```
Comment (wehrli) [inline Functions/src/views/index.html:58]:
```suggestion
              <button class="btn btn-outline-primary btn-sm" type="button" id="checkLength">Length</button>
```

File: Functions/src/views/index.html [hunk 18]
```diff
@@ -11,43 +11,128 @@
...
+            <p id="length"></p>
+          </div>
+        </li>
+
+        <!-- Exercice 5 -->
+        <li>
+          <div class="input-group mt-2">
+            <input type="text" id="inputArray" class="form-control pr-5" placeholder='Ex: ["test", "test", "test"]' aria-label="Length of array" aria-describedby="Length of array">
+            <div class="input-group-prepend mr-2">
+              <button class="btn btn-outline-primary btn-sm" type="button" id="lengthArray">Lenght of array</button>
```
Comment (wehrli) [inline Functions/src/views/index.html:71]:
```suggestion
              <button class="btn btn-outline-primary btn-sm" type="button" id="lengthArray">Length of array</button>
```

**Review by wehrli — COMMENTED** [review]
way more better, nice job !

remarks: 
- On the index html put a title or a description per exercice (I know each one of them, but think of the fact that a lambda user could use your app)
- You need to give a better naming for each value you are creating. string2, button1, input4 that's not enough representative of what this variable id for.

**Review by wehrli — APPROVED** [review]

**Comment by wehrli** [comment]
you can fix the things I mentioned if you feel like :), but don't forget to merge what you've done with master

File: Functions/src/js/func.js (hunk 11 above)
Comment (geoffrey112) [reply Functions/src/js/func.js:107]: Because i only want the int 🧐

File: Functions/src/js/func.js (hunk 12 above)
Comment (geoffrey112) [reply Functions/src/js/func.js:117]: Oups, Okey 👍

File: Functions/src/js/func.js (hunk 14 above)
Comment (geoffrey112) [reply Functions/src/js/func.js:154]: My bad🤦‍♂️

File: Functions/src/js/func.js (hunk 8 above)
Comment (geoffrey112) [reply Functions/src/js/func.js:170]: Okey 🙂👍

File: Functions/src/js/func.js (hunk 11 above)
Comment (wehrli) [reply Functions/src/js/func.js:107]: okay but here you are blocking only the letter `e` so it is not bullet proof ^^. You want to use regexp here

**Pull request merged**